README.md
# Turaco [![Build Status](https://travis-ci.org/mitchhentges/turaco.svg?branch=master)](https://travis-ci.org/mitchhentges/turaco/)

Lightweight boolean expression reducer, originally called [BExpred](https://sourceforge.net/projects/bexpred/).
Compatible with Java, all the way back to 1.3.

* Written by: Benjamin Biron bbiron@users.sourceforge.net
* Modified by: Mitchell Hentges mitchhentges@protonmail.com

![](doc-resources/screenshot.png)

## Instructions

1) To run it:
	a) Unzip / untar the downloaded file
	b) In a console / command prompt, go to the "jar" directory
	c) From there: java -jar turaco.jar
	NOTE: You'll need to have the java runtime in your PATH, for windows
	users, it's somewhere in the System icon from the control panel (May be
	called "Environment Variables"). For linux users running bash, it's
	likely in your ~/.bash_profile file.
	
2) Entering an expression:
	A valid boolean expression must be entered ex: A+B, A*B+C Var1+Var2^Var3
	The variable names may consist of anything except spaces and operator
	chars (+, *, !, etc).

	Operators and corresponding chars:
		AND	* or &
		OR	+ or |
		XOR	^
		NOT	! or ~

	You MUST specify an operator between 2 variables:
		A*B is NOT the same as AB

	If no operator is found before an opening bracket or after a closing
	one, an AND is assumed:
		(A+B)(C+A) is the same as (A+B)*(C+A)

3) Reducing the expression:
	The expression will be reduced if you press ENTER in the expression
	field or click the "Reduce" button. The SoP and PoS checkboxes specify
	whether you want the expression to be reduced as "Sum of Products" or
	"Product of Sums" or both.
	
	Expression with a large number of variables may take time to reduce and
	this is why you have the option to limit the reduction process by
	selecting only one form.

4) Evaluating an expression with given values:
	You may want to only evaluate an expression with certain values. This is
	achieved by typing in the expression, pressing enter, checking the
	variables you want as "1" and finally clicking the "Evaluate" button.

5) Other tools:
	There are other handy tools provided with Turaco:
	a) Truth table:
		Shows the truth table corresponding to the expression currently
		typed in the expression field.
		
	b) Expression comparator:
		Allows you to compare 2 expressions for equality

## Notes

The expressions are reduced via the Quine and Mc.Cluskey method. You can select
whether to output the reduced expression in the "Sum of Products (SoP)" form, the
"Product of Sums (PoS)" or both.

In some cases, they're may be more than 1 equivalently reduced valid expression,
eventually, I'll make sure to display all possibilities, but for now, only one
is displayed.

A minor code cleanup and commenting session is due, I'll try to make the time
for it soon.

## Development

1. Initialize your IDE in the `turaco` directory
2. Run the `main()` method in `BExpred.java` to launch `Turaco`

The files can be jar'd with the `jar.sh` script

## Why is this called Turaco?

Due to the GNU GPL v2, I can't use the name BExpred anymore. [Birbs](http://birdcheese.tumblr.com/)
are cool, though, and turacos look really neat.

![](doc-resources/turaco.jpg)

## License

Released under the GNU GPL, see LICENSE.txt for details.


HISTORY.txt
** Version 1.1.0 - 2016/05/10
- Flag field as "needs refresh" after paste
- A _lot_ of cleanup. There's a solid potential for regressions.

** Version 1.0.2 - 2016/05/01
- Fixed another case where an unnecessary StackOverflow occurred

** Version 1.0.1 - 2016/05/01
- Clear exception text box after successful evaluation
- Don't evaluate expression on focus-lost if still showing placeholder text

** Version 1.0.0 - 2016/04/29
- Clear "Enter an expression" when a character is entered
- Show the current error in a single-line TextBox on the bottom of the panel
- Fixed a case where StackOverflowError would be thrown on a short string, such as: `(B)*((B)*(B))*((B)*(B))`
    This is because BExprParser.nextOp() wasn't considering bracket depth

** Version 0.8 - 2003/06/07
- Finally got around to fixing the little annoying parsing bugs for expressions
  such as the following:
	!A
	!!A
	!(!A)
	(((A)))
	(A + B) + (C + D) <Note the extra spaces>

  Thanks to Hanh-Missi for pointing them out.

** Version 0.7 - 2003/04/19
- Minor bugfixes in the reduction. Thanks to Jan for pointing them out.

** Version 0.6 - 2003/03/16
- Now reduces!
- Reduced output in "sum of products" and "product of sums" form
- Better truth table

** Version 0.5 - 2003/03/06
- Added a menu bar
- Now includes an expression comparator!
- Minor bug fixes

** Version 0.4 - 2003/02/18
- Support for NOTs, use ! or ~
- Fixed various bugs relating to operator precedence and brackets
- Added primitive truth table (A better one will come)
- Replaced input TextFields with checkboxes
- Added a ScrollPane for the checkboxes
- Expression tree will auto-refresh when the expression changed and focus is
  lost in the expression input field
- You can now press enter in the expression input field to refresh the tree

** Version 0.3 - 2003/01/29
- Completely rewritten expression parser
- Support for XORs, use ^

** Version 0.2 - 2003/01/26
- Initial version


src/BExpredMain.java
package bexpred;

import java.awt.*;
import java.awt.event.*;
import javax.swing.*;

import javax.swing.table.TableColumn;
import javax.swing.table.TableCellEditor;

import java.util.Hashtable;
import java.util.ArrayList;
import javax.swing.event.*;

/*
 * BExpred - Boolean Expression Reducer
 * Goal: To reduce specified boolean expressions
 * Current state: Evaluates a given expression with input values
 * Copyright (c) 2003 Benjamin Biron
 * License: GPL

    This file is part of BExpred.

    BExpred is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    BExpred is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with BExpred; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

public class BExpredMain extends JFrame {
  JPanel jPanel1 = new JPanel();
  JTextField ExprField = new JTextField();
  BorderLayout borderLayout1 = new BorderLayout();
  JButton EvaluateBtn = new JButton();
  TTFrame ttFrame;
  BExprCompare cmpFrame;

  BExprTree bexprtree = null;
  ArrayList var_inputs = new ArrayList();
  JTextField valueField = new JTextField();
  JButton TTBtn = new JButton();


  boolean needs_refresh = false;
  JMenuBar jMenuBar1 = new JMenuBar();
  JMenu jMenu1 = new JMenu();
  JMenuItem jMenuItem1 = new JMenuItem();
  JMenu jMenu2 = new JMenu();
  JCheckBoxMenuItem jCheckBoxMenuItem1 = new JCheckBoxMenuItem();
  JCheckBoxMenuItem jCheckBoxMenuItem2 = new JCheckBoxMenuItem();
  JButton ReduceBtn = new JButton();
  JLabel jLabel1 = new JLabel();
  JLabel jLabel2 = new JLabel();
  JTextField SOPField = new JTextField();
  JPanel jPanel3 = new JPanel();
  JTextField POSField = new JTextField();
  JScrollPane jScrollPane1 = new JScrollPane();
  JPanel jPanel2 = new JPanel();
  JCheckBox SOPCheck = new JCheckBox();
  JCheckBox POSCheck = new JCheckBox();
  //Construct the frame
  public BExpredMain() {
    enableEvents(AWTEvent.WINDOW_EVENT_MASK);
    try {
      jbInit();
    }
    catch(Exception e) {
      e.printStackTrace();
    }
  }
  //Component initialization
  private void jbInit() throws Exception  {
    //this.setContentPane(jPanel1);
    this.setDefaultCloseOperation(HIDE_ON_CLOSE);
    this.setEnabled(true);
    this.setForeground(Color.black);
    this.setLocale(java.util.Locale.getDefault());
    this.setJMenuBar(jMenuBar1);
    this.setResizable(false);
    this.setSize(new Dimension(392, 385));
    this.setState(Frame.NORMAL);
    this.setTitle("BExpred");
    ExprField.setSelectionStart(19);
    ExprField.setText("Enter an expression");
    ExprField.setBounds(new Rectangle(7, 6, 378, 19));
    ExprField.addFocusListener(new BExpredMain_ExprField_focusAdapter(this));
    ExprField.addKeyListener(new BExpredMain_ExprField_keyAdapter(this));
    ExprField.addMouseListener(new BExpredMain_ExprField_mouseAdapter(this));
    jPanel1.setAlignmentX((float) 0.5);
    jPanel1.setAlignmentY((float) 0.5);
    jPanel1.setOpaque(true);
    jPanel1.setLayout(null);
    EvaluateBtn.setBounds(new Rectangle(236, 352, 86, 25));
    EvaluateBtn.setText("Evaluate");
    EvaluateBtn.addActionListener(new BExpredMain_EvaluateBtn_actionAdapter(this));
    valueField.setText("");
    valueField.setBounds(new Rectangle(326, 355, 56, 19));
    ttFrame = new TTFrame();
    TTBtn.setBounds(new Rectangle(278, 30, 107, 25));
    TTBtn.setText("Truth Table");
    TTBtn.addActionListener(new BExpredMain_TTBtn_actionAdapter(this));
    cmpFrame = new BExprCompare();
    cmpFrame.setJMenuBar(null);
    jMenu1.setText("File");
    jMenuItem1.setText("Exit");
    jMenuItem1.addMouseListener(new BExpredMain_jMenuItem1_mouseAdapter(this));
    jMenu2.setText("Tools");
    jMenu2.addMouseListener(new BExpredMain_jMenu2_mouseAdapter(this));
    jCheckBoxMenuItem1.setText("Truth Table");
    jCheckBoxMenuItem1.addMouseListener(new BExpredMain_jCheckBoxMenuItem1_mouseAdapter(this));
    jCheckBoxMenuItem2.setText("Expression comparator");
    jCheckBoxMenuItem2.addMouseListener(new BExpredMain_jCheckBoxMenuItem2_mouseAdapter(this));
    ReduceBtn.setBounds(new Rectangle(7, 30, 88, 25));
    ReduceBtn.setActionCommand("Reduce");
    ReduceBtn.setText("Reduce");
    ReduceBtn.addActionListener(new BExpredMain_ReduceBtn_actionAdapter(this));
    jLabel1.setFont(new java.awt.Font("SansSerif", 0, 12));
    jLabel1.setToolTipText("");
    jLabel1.setText("Sum of Products");
    jLabel1.setBounds(new Rectangle(9, 64, 98, 15));
    jLabel2.setFont(new java.awt.Font("SansSerif", 0, 12));
    jLabel2.setText("Product of Sums");
    jLabel2.setBounds(new Rectangle(9, 83, 98, 16));
    SOPField.setText("");
    SOPField.setBounds(new Rectangle(111, 62, 274, 19));
    jPanel3.setLayout(null);
    jPanel3.setAlignmentY((float) 0.5);
    jPanel3.setAlignmentX((float) 0.5);
    POSField.setText("");
    POSField.setBounds(new Rectangle(111, 82, 274, 19));
    jScrollPane1.setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_ALWAYS);
    jScrollPane1.setVerticalScrollBarPolicy(JScrollPane.VERTICAL_SCROLLBAR_NEVER);
    jScrollPane1.setBounds(new Rectangle(9, 113, 373, 228));
    jPanel2.setBounds(new Rectangle(3, 3, 102, 80));
    jPanel2.setLayout(null);
    SOPCheck.setSelected(true);
    SOPCheck.setText("SoP");
    SOPCheck.setBounds(new Rectangle(96, 32, 50, 23));
    POSCheck.setSelected(true);
    POSCheck.setText("PoS");
    POSCheck.setBounds(new Rectangle(145, 32, 49, 23));
    jPanel1.add(ExprField, null);
    jPanel1.add(ReduceBtn, null);
    jPanel1.add(TTBtn, null);
    jPanel1.add(jLabel1, null);
    jPanel1.add(POSField, null);
    jPanel1.add(SOPField, null);
    jPanel1.add(jLabel2, null);
    jPanel1.add(jScrollPane1, null);
    jPanel1.add(EvaluateBtn, null);
    jPanel1.add(valueField, null);
    jPanel1.add(SOPCheck, null);
    jScrollPane1.getViewport().add(jPanel2, null);
    jMenuBar1.add(jMenu1);
    jMenuBar1.add(jMenu2);
    jMenu1.add(jMenuItem1);
    jMenu2.add(jCheckBoxMenuItem1);
    jMenu2.add(jCheckBoxMenuItem2);
    this.getContentPane().add(jPanel1, BorderLayout.NORTH);
    jPanel1.add(POSCheck, null);

    //Adjusting the size of the main window
    Dimension goodSize = this.jScrollPane1.getPreferredSize();
    goodSize.width += 2*(this.jScrollPane1.getX());
    goodSize.height += this.jScrollPane1.getY() + 10;
    jPanel1.setPreferredSize(new Dimension(390, 385));
  }
  //Overridden so we can exit when window is closed
  protected void processWindowEvent(WindowEvent e) {
    super.processWindowEvent(e);
    if (e.getID() == WindowEvent.WINDOW_CLOSING) {
      System.exit(0);
    }
  }

  void refreshTree() {
    JCheckBox aCheck;
    Dimension dim;
    int col, lastCol, colPos, maxColWidth;

    this.needs_refresh = false;

    // Create the expression tree from the expression.
    try {
      this.bexprtree = new BExprTree(this.ExprField.getText());
      //System.out.println(this.bexprtree.getTruthTable().getExpr(this.bexprtree.getTruthTable().reduceVars(this.bexprtree.getVars())));
    } catch (BExprPreParseException ex) {
      this.bexprtree = null;
      // Need a dialog!!
      ex.printStackTrace();
    }

    // Must clear the panel of the labels and input fields
    for (int i = 0; i < var_inputs.size(); i++) {
      this.jPanel2.remove((JCheckBox) var_inputs.get(i));
      //this.jScrollPane1.getViewport().remove((JCheckBox) var_inputs.get(i));
    }

    this.var_inputs.clear();

    if (this.bexprtree != null) {
      // The list of vars in order of appearance
      ArrayList vars = this.bexprtree.getVars();

      // Let's now create the input fields and labels
      col = 0;
      lastCol = 0;
      colPos = 5;
      maxColWidth = 0;
      for (int i = 0; i < vars.size(); i++) {
        col = (i/9);
        if (col != lastCol) {
          colPos += maxColWidth + 10;
          maxColWidth = 0;
        }

        aCheck = new JCheckBox((String) vars.get(i));
        dim = aCheck.getPreferredSize();

        if ((int) dim.getWidth() > maxColWidth) {
          maxColWidth = (int) dim.getWidth();
        }

        //aCheck.setBounds(new Rectangle((i/9)*100 + 32, (i%9)*22 + 67, 68, 19));
        aCheck.setBounds(new Rectangle(colPos, (i%9)*22 + 5, (int) dim.getWidth(), (int) dim.getHeight()));
        this.jPanel2.add(aCheck, null);
        //this.jScrollPane1.getViewport().add(aCheck, null);
        this.var_inputs.add(aCheck);
        lastCol = col;
      }

      this.jPanel2.setPreferredSize(new Dimension(colPos + maxColWidth, this.jPanel2.getHeight()));
      this.jPanel2.updateUI();
      //this.jScrollPane1.getViewport().setPreferredSize(new Dimension(colPos + maxColWidth, this.jScrollPane1.getViewport().getHeight()));
      //this.jScrollPane1.getViewport().updateUI();

      this.updateTT(this.bexprtree.getTruthTable());
      //TESTING System.out.println(this.bexprtree.getTruthTable().getExpr(this.bexprtree.getVars()));
    }
  }

  void goReduce() {
    if (this.SOPCheck.isSelected())
      this.SOPField.setText(this.bexprtree.getTruthTable().getSOP(this.bexprtree.getVars()));
    else
      this.SOPField.setText("Option unchecked");

    if (this.POSCheck.isSelected())
      this.POSField.setText(this.bexprtree.getTruthTable().getPOS(this.bexprtree.getVars()));
    else
      this.POSField.setText("Option unchecked");
  }

  void updateTT(TruthTable aTT) {
    ArrayList vars = this.bexprtree.getVars();
    String[] colNames = new String[vars.size() + 1];

    for (int i = 0; i < vars.size(); i++) {
      colNames[i] = (String) vars.get(i);
    }

    colNames[vars.size()] = "Output";

    this.ttFrame.jScrollPane1.getViewport().removeAll();
    //this.ttFrame.jTable1 = new JTable(aTT.getWrappedTable(true), colNames);
    this.ttFrame.jTable1 = new JTTable(aTT.getWrappedTable(), colNames);
    this.ttFrame.jScrollPane1.getViewport().add(this.ttFrame.jTable1);

    // Set the dimensions of the TTFrame
    Dimension screenSize = Toolkit.getDefaultToolkit().getScreenSize();
    int ttHeight = (int) this.ttFrame.jTable1.getPreferredSize().getHeight() + 44;
    ttHeight = screenSize.getHeight() - 60 < ttHeight ? ((int) screenSize.getHeight() - 60) : ttHeight;
    this.ttFrame.setSize(this.ttFrame.jTable1.getColumnCount() * 60, ttHeight);
  }

  void ExprTreeBtn_actionPerformed(ActionEvent e) {
    this.refreshTree();
  }

  void EvaluateBtn_actionPerformed(ActionEvent e) {
    Hashtable values = new Hashtable();
    ArrayList vars = this.bexprtree.getVars();

    // Evaluate the expression by passing a HashTable of VAR:VALUE, if VALUE is
    // anything other than !, it will be taken as false.
    for (int i = 0; i < var_inputs.size(); i++) {
      values.put((String) vars.get(i), ((JCheckBox) var_inputs.get(i)).isSelected() ? "1" : "0");
    }
    //DEBUG System.out.println(bexprtree.evaluate(values));
    this.valueField.setText(bexprtree.evaluate(values) ? "True" : "False");
  }

  void ExprField_mouseClicked(MouseEvent e) {
    // This just clears the ExprField when it's clicked and contains "Enter an expression"
    if (this.ExprField.getText().compareTo("Enter an expression") == 0) {
      this.ExprField.setText("");
    }
  }

  void TTBtn_actionPerformed(ActionEvent e) {
    //Bah, just pops up what's going to be the truth table
    if (this.bexprtree != null) {
      this.ttFrame.show();
    } else {
      // Should be a dialog
      System.out.println("There are no expressions to display in the truth table");
    }

  }

  void ExprField_keyPressed(KeyEvent e) {
    if (e.getKeyCode() == 10) {
      this.refreshTree();
      this.goReduce();
    } else if (!e.isAltDown() && !e.isControlDown()) {
      this.needs_refresh = true;
    }
  }

  void ExprField_focusLost(FocusEvent e) {
    if (this.needs_refresh)
      this.refreshTree();
  }

  void jMenuItem1_mouseReleased(MouseEvent e) {
    System.exit(0);
  }

  void jMenu2_mousePressed(MouseEvent e) {
    if (this.ttFrame.isVisible())
      jCheckBoxMenuItem1.setState(true);
    else
      jCheckBoxMenuItem1.setState(false);

    if (this.cmpFrame.isVisible())
      jCheckBoxMenuItem2.setState(true);
    else
      jCheckBoxMenuItem2.setState(false);
  }

  void jCheckBoxMenuItem1_mouseReleased(MouseEvent e) {
    if (jCheckBoxMenuItem1.getState()) {
      if (this.bexprtree != null)
        this.ttFrame.show();
      else
        System.out.println("There are no expressions to display in the truth table");
    } else {
      this.ttFrame.hide();
    }
  }

  void jCheckBoxMenuItem2_mouseReleased(MouseEvent e) {
    if (jCheckBoxMenuItem2.getState()) {
      this.cmpFrame.pack();
      this.cmpFrame.show();
    } else {
      this.cmpFrame.hide();
    }
  }

  void ReduceBtn_actionPerformed(ActionEvent e) {
    this.goReduce();
  }
}

class BExpredMain_EvaluateBtn_actionAdapter implements java.awt.event.ActionListener {
  BExpredMain adaptee;

  BExpredMain_EvaluateBtn_actionAdapter(BExpredMain adaptee) {
    this.adaptee = adaptee;
  }
  public void actionPerformed(ActionEvent e) {
    adaptee.EvaluateBtn_actionPerformed(e);
  }
}

class BExpredMain_ExprField_mouseAdapter extends java.awt.event.MouseAdapter {
  BExpredMain adaptee;

  BExpredMain_ExprField_mouseAdapter(BExpredMain adaptee) {
    this.adaptee = adaptee;
  }
  public void mouseClicked(MouseEvent e) {
    adaptee.ExprField_mouseClicked(e);
  }
}

class BExpredMain_TTBtn_actionAdapter implements java.awt.event.ActionListener {
  BExpredMain adaptee;

  BExpredMain_TTBtn_actionAdapter(BExpredMain adaptee) {
    this.adaptee = adaptee;
  }
  public void actionPerformed(ActionEvent e) {
    adaptee.TTBtn_actionPerformed(e);
  }
}

class BExpredMain_ExprField_keyAdapter extends java.awt.event.KeyAdapter {
  BExpredMain adaptee;

  BExpredMain_ExprField_keyAdapter(BExpredMain adaptee) {
    this.adaptee = adaptee;
  }
  public void keyPressed(KeyEvent e) {
    adaptee.ExprField_keyPressed(e);
  }
}

class BExpredMain_ExprField_focusAdapter extends java.awt.event.FocusAdapter {
  BExpredMain adaptee;

  BExpredMain_ExprField_focusAdapter(BExpredMain adaptee) {
    this.adaptee = adaptee;
  }
  public void focusLost(FocusEvent e) {
    adaptee.ExprField_focusLost(e);
  }
}

class BExpredMain_jMenuItem1_mouseAdapter extends java.awt.event.MouseAdapter {
  BExpredMain adaptee;

  BExpredMain_jMenuItem1_mouseAdapter(BExpredMain adaptee) {
    this.adaptee = adaptee;
  }
  public void mouseReleased(MouseEvent e) {
    adaptee.jMenuItem1_mouseReleased(e);
  }
}

class BExpredMain_jMenu2_mouseAdapter extends java.awt.event.MouseAdapter {
  BExpredMain adaptee;

  BExpredMain_jMenu2_mouseAdapter(BExpredMain adaptee) {
    this.adaptee = adaptee;
  }
  public void mousePressed(MouseEvent e) {
    adaptee.jMenu2_mousePressed(e);
  }
}

class BExpredMain_jCheckBoxMenuItem1_mouseAdapter extends java.awt.event.MouseAdapter {
  BExpredMain adaptee;

  BExpredMain_jCheckBoxMenuItem1_mouseAdapter(BExpredMain adaptee) {
    this.adaptee = adaptee;
  }
  public void mouseReleased(MouseEvent e) {
    adaptee.jCheckBoxMenuItem1_mouseReleased(e);
  }
}

class BExpredMain_jCheckBoxMenuItem2_mouseAdapter extends java.awt.event.MouseAdapter {
  BExpredMain adaptee;

  BExpredMain_jCheckBoxMenuItem2_mouseAdapter(BExpredMain adaptee) {
    this.adaptee = adaptee;
  }
  public void mouseReleased(MouseEvent e) {
    adaptee.jCheckBoxMenuItem2_mouseReleased(e);
  }
}

class BExpredMain_ReduceBtn_actionAdapter implements java.awt.event.ActionListener {
  BExpredMain adaptee;

  BExpredMain_ReduceBtn_actionAdapter(BExpredMain adaptee) {
    this.adaptee = adaptee;
  }
  public void actionPerformed(ActionEvent e) {
    adaptee.ReduceBtn_actionPerformed(e);
  }
}